File: storage/reading.rb
require 'time'
require_relative 'device_and_reading_storage'


class Reading
  attr_reader :id, :device_id, :timestamp
  attr_accessor :count


  def initialize(device_id:, timestamp:, count:)
    @id = nil
    @device_id = device_id
    @timestamp = timestamp.is_a?(Time) ? timestamp : Time.parse(timestamp)
    @count = count.to_i
  end


  def save!
    unless @id
      DeviceAndReadingStorage.instance.add_reading(self)
    end
    self
  end


  def persisted?
    !@id.nil?
  end

end

File: express_api/src/reading.js
const storage = require('./storage');


class Reading {

  constructor(deviceId, timestamp, count) {
    this.id = null;
    this.deviceId = deviceId;
    this.timestamp = timestamp instanceof Date ? timestamp : new Date(timestamp);
    this.count = parseInt(count, 10);
  }


  save() {
    if (!this.id) {
      storage.addReading(this);
    }
    return this;
  }

}


module.exports = Reading;
